### ISSUE TITLE
wxDialog disappears when another is destroyed

### ISSUE THREAD

**Issue by nietsu** [wxWidgets/wxWidgets#23560]
### Description

When two wxDialogs are created and shown and then the first one is destroyed, the other dialog also disappears. This will not happen if both dialogs are created before either are shown, or if the dialogs have a parent.

For example, in this code sample the dialogs will show quickly and then both disappear:
```
auto f1 = new wxDialog(nullptr, wxID_ANY, "D1");
f1->Show();
auto f2 = new wxDialog(nullptr, wxID_ANY, "D2");
f2->Show();
f1->Destroy(); // This will also make f2 disappear unexpectedly
```
`f2` will only disappear. It is not destroyed completely, and its `IsShown()` still returns `true`.
In this sample only the first dialog will disappear and the second will remain on screen, as it should. This is the expected behaviour:
```
auto f1 = new wxDialog(nullptr, wxID_ANY, "D1");
auto f2 = new wxDialog(nullptr, wxID_ANY, "D2");
f1->Show();
f2->Show();
f1->Destroy(); // Destroys only f1. f2 will remain visible as it should
```

Here's a patch to dialogs sample to cause the same issue. Two dialogs are created on launch and the first one destroyed.
Expected behaviour would be for the second dialog to remain visible.
[disappearingdialog.patch](https://github.com/wxWidgets/wxWidgets/files/11542591/disappearingdialog.patch)

### Platform and version information

- wxWidgets version  you use: 3.2.2.1, also current master 563cc11151ee0d986dae8939c4bb24a3a42c0611
- wxWidgets port  you use: wxMSW
- OS  and its version: Windows 11 Pro 22H2 22621.1702

**Comment by vadz** [comment]
This is a consequence of implicitly using D1 as the parent of D2. Because unowned dialogs are usually a bad thing, wxWidgets tries to find the correct owner automatically if none is specified.

If you really don't want to have any owner for the second dialog, you can use `wxDIALOG_NO_PARENT`:
```cpp
    auto f2 = new wxDialog(nullptr, wxID_ANY, "D2", wxDefaultPosition, wxDefaultSize, wxDEFAULT_DIALOG_STYLE | wxDIALOG_NO_PARENT);
```

If `f1` is really temporary, it's probably best to create it as a transient window:
```cpp
    auto f1 = new wxDialog(nullptr, wxID_ANY, "D1");
    f1->SetExtraStyle(wxWS_EX_TRANSIENT);
```
as this lets `f2` be owned by the main frame. Of course, if this is really what you want, you can just pass it instead of `nullptr` as the parent.

Finally, the most general advice is to never use non-modal dialogs at all.

Of course, all this being said, it's arguably still a bug that this happens in the first place. But I don't know how to fix it without losing useful functionality.

**Comment by nietsu** [comment]
Thank you for the explanation. I had totally missed the `wxDIALOG_NO_PARENT` flag! It might actually be cool if `wxDialog` constructor checked with an assert that either parent or `wxDIALOG_NO_PARENT` flag is set.

Seems that `GetParent()` still returns `nullptr` without the flag, which might explain why the `wxDialog` object does not get destroyed with its implicitly set parent.

**Comment by vadz** [comment]
> Thank you for the explanation. I had totally missed the `wxDIALOG_NO_PARENT` flag! It might actually be cool if `wxDialog` constructor checked with an assert that either parent or `wxDIALOG_NO_PARENT` flag is set.

No, it's intentionally not done like this -- null parent doesn't mean "no parent" (if it did, we wouldn't need `wxDIALOG_NO_PARENT`) because it has historically always been used as "choose the appropriate parent automatically". I.e. since the dawn of time (meaning 1993 or whenever) people called `wxMessageBox()` without specifying the parent explicitly but still expected it to behave correctly -- which is only the case under MSW if it uses the correct parent, i.e. usually the main application window. And this is really useful, having to manually specify the parent all the time is tiresome.

Of course, this logic only works for the modal dialogs. But the trouble is that we have to set the parent before we know if the dialog is going to be used modally or not (well, we could change the parent later in `ShowModal()`, I guess...). Which is why I really wish we had never had non-modal dialogs in the first place.

**Comment by tanzislam** [comment]
I'd recommend not changing the existing behaviour at this point, but instead documenting the meaning of `nullptr` parent in the `wxDialog` constructor documentation, in addition to in the `wxDIALOG_NO_PARENT` style flag's documentation. (I had also assumed `nullptr` parent means an orphan/top-level dialog, similar to what happens with `wxFrame`.)

**Comment by vadz** [comment]
Good point, I'll add a note about this.
